nixio/cmd/upgrade.py

```diff
@@ -0,0 +1,219 @@
+
+        def find_props(name, group):
+            if isinstance(group, h5py.Dataset) and len(group.dtype):
+                # structured/compound dtypes have non-zero length
+                props.append(group.name)
+
+        sections.visititems(find_props)
+
+    if not props:
+        return None
+
+    def update_props():
+        for propname in props:
+            with h5py.File(fname, mode="a") as hfile:
+                prop = hfile[propname]
+                if not (isinstance(prop, h5py.Dataset) and len(prop.dtype)):
+                    # File was possibly changed since the tasks were
+                    # collected.  File may have been submitted twice or
+                    # multiple instances of the script could be running.
+                    # skip this prop
+                    continue
+
+                print(f"Fixing {propname}")
+                # pull out the old extra attributes
+                uncertainty = prop["uncertainty"]
+                reference = prop["reference"]
+                filename = prop["filename"]
+                encoder = prop["encoder"]
+                checksum = prop["checksum"]
+
+                # replace base prop
+                values = prop["value"]
+                dt = values.dtype
+                del hfile[propname]
+                newprop = create_property(hfile, propname,
+                                          dtype=dt, data=values)
+
+                # Create properties for any extra attrs that are set
+                if len(set(uncertainty)) > 1:
+                    # multiple values, make new prop
+                    create_property(hfile, propname + ".uncertainty",
+                                    dtype=float, data=uncertainty)
+                elif any(uncertainty):
+                    # single, unique, non-zero value; add to main prop attr
+                    newprop.attrs["uncertainty"] = uncertainty[0]
+
+                if any(reference):
+                    create_property(hfile, propname + ".reference",
+                                    dtype=nix.util.vlen_str_dtype,
+                                    data=reference)
+
+                if any(filename):
+                    create_property(hfile, propname + ".filename",
+                                    dtype=nix.util.vlen_str_dtype,
+                                    data=filename)
+
+                if any(encoder):
+                    create_property(hfile, propname + ".encoder",
+                                    dtype=nix.util.vlen_str_dtype,
+                                    data=encoder)
+
+                if any(checksum):
+                    create_property(hfile, propname + ".checksum",
+                                    dtype=nix.util.vlen_str_dtype,
+                                    data=checksum)
+
+    update_props.__doc__ = "Update {} properties".format(len(props))
+    return update_props
+
+
+def create_property(hfile, name, dtype, data):
```

Comment:
I am afraid the unit got lost during upgrading

Reply:
Oh dear.  Good catch!